library/core/tests/mem.rs

```diff
@@ -129,3 +129,20 @@ fn test_discriminant_send_sync() {
     is_send_sync::<Discriminant<Regular>>();
     is_send_sync::<Discriminant<NotSendSync>>();
 }
+
+#[test]
+fn assume_init_good() {
+    const TRUE: bool = unsafe { MaybeUninit::<bool>::new(true).assume_init() };
+
+    assert!(TRUE);
+}
+
+#[test]
+fn assume_init_bad() {
+    const _BAD: () = unsafe {
+        MaybeUninit::<!>::uninit().assume_init();
+        //~^ ERROR the type `!` does not permit being left uninitialized
+        //~| ERROR this code causes undefined behavior when executed
+        //~| ERROR help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done
```

Comment:
I am having some trouble getting the tests to compile [gist](https://gist.github.com/usbalbin/f4e58790c85ebf3f7ac04b5151c37aa4). First off is 
```
`"calling intrinsic `assert_inhabited`" needs an rfc before being allowed inside constants`. 
```
in both tests where I am calling `MaybeUninit::assume_init`. I have added both the feature gate for `assume_init` and the gate for the intrinsic, still the same error.

---

The other error is
```
error: the type `!` does not permit being left uninitialized
```
which is what to expect when writing that code. But how do I make the test expect that error. I have read, and tried to follow the [Dev guide](https://rustc-dev-guide.rust-lang.org/tests/adding.html) but it does not seem to work.

Reply:
You're running `x.py test --stage 0 library/core`. You need to use `x.py test --stage 1` to include your changes.

Reply:
Also the tests may need `cfg(not(bootstrap))`.

Reply:
Thanks a lot

Reply:
When running the test locally with `x.py test --stage 1 library/core` I only get 
```
`assert_inhabited`" needs an rfc`
```
for the `assume_init_bad` test as well as the 
```
`the type `!` does not permit being left uninitialized`
```
The `assert_inhabited` error for the `assume_init_good` test (which is visible on [github CI](https://github.com/rust-lang/rust/pull/79621/checks?check_run_id=1504223786)) does not show.. And when I completely comment out the `assume_init_bad` test, all errors disappear.

Reply:
`//~ ERROR` annotations don't work in unit tests, they only work in compiletest (`src/test/ui`).

What are you even trying to test here? If this is about making CTFE fail and testing the error, you need to create a compile-fail test by creating a new file such as `src/test/ui/consts/assume-type-intrinsics.rs`.

Reply:
Yes, I was trying to implement the tests you suggested in [your comment](https://github.com/rust-lang/rust/pull/79621#issuecomment-737129002). So the "BAD" test should go into `src/test/ui` then, that makes sense :)